mono/profiler/coverage.c

```diff
@@ -0,0 +1,923 @@
+	mono_conc_hashtable_insert (coverage_profiler.methods, method, method);
+	mono_conc_hashtable_insert (coverage_profiler.assemblies, assembly, assembly);
+	mono_os_mutex_unlock (&coverage_profiler.mutex);
+
+	image_methods = (MonoLockFreeQueue *)mono_conc_hashtable_lookup (coverage_profiler.image_to_methods, image);
+
+	if (image_methods == NULL) {
+		image_methods = (MonoLockFreeQueue *) g_malloc (sizeof (MonoLockFreeQueue));
+		mono_lock_free_queue_init (image_methods);
+		mono_os_mutex_lock (&coverage_profiler.mutex);
+		mono_conc_hashtable_insert (coverage_profiler.image_to_methods, image, image_methods);
+		mono_os_mutex_unlock (&coverage_profiler.mutex);
+	}
+
+	node = create_method_node (method);
+	mono_lock_free_queue_enqueue (image_methods, node);
+
+	class_methods = (MonoLockFreeQueue *)mono_conc_hashtable_lookup (coverage_profiler.classes, klass);
+
+	if (class_methods == NULL) {
+		class_methods = (MonoLockFreeQueue *) g_malloc (sizeof (MonoLockFreeQueue));
+		mono_lock_free_queue_init (class_methods);
+		mono_os_mutex_lock (&coverage_profiler.mutex);
+		mono_conc_hashtable_insert (coverage_profiler.classes, klass, class_methods);
+		mono_os_mutex_unlock (&coverage_profiler.mutex);
+	}
+
+	node = create_method_node (method);
+	mono_lock_free_queue_enqueue (class_methods, node);
+
+	return TRUE;
+}
+
+#define LINE_BUFFER_SIZE 4096
+/* Max file limit of 128KB */
+#define MAX_FILE_SIZE 128 * 1024
+static char *
+get_file_content (const gchar *filename)
+{
+	char *buffer;
+	ssize_t bytes_read;
+	long filesize;
+	int res, offset = 0;
+	FILE *stream;
+
+	stream = fopen (filename, "r");
+	if (stream == NULL)
+		return NULL;
+
+	res = fseek (stream, 0, SEEK_END);
+	if (res < 0) {
+		fclose (stream);
+		return NULL;
+	}
+
+	filesize = ftell (stream);
+	if (filesize < 0) {
+		fclose (stream);
+		return NULL;
+	}
+
+	res = fseek (stream, 0, SEEK_SET);
+	if (res < 0) {
+		fclose (stream);
+		return NULL;
+	}
+
+	if (filesize > MAX_FILE_SIZE) {
+		fclose (stream);
+		return NULL;
+	}
+
+	buffer = (char *) g_malloc ((filesize + 1) * sizeof (char));
+	while ((bytes_read = fread (buffer + offset, 1, LINE_BUFFER_SIZE, stream)) > 0)
+		offset += bytes_read;
+
+	/* NULL terminate our buffer */
+	buffer[filesize] = '\0';
+
+	fclose (stream);
+	return buffer;
+}
+
+static char *
+get_next_line (char *contents, char **next_start)
+{
+	char *p = contents;
+
+	if (p == NULL || *p == '\0') {
+		*next_start = NULL;
+		return NULL;
+	}
+
+	while (*p != '\n' && *p != '\0')
+		p++;
+
+	if (*p == '\n') {
+		*p = '\0';
+		*next_start = p + 1;
+	} else
+		*next_start = NULL;
+
+	return contents;
+}
+
+static void
+init_suppressed_assemblies (void)
+{
+	char *content;
+	char *line;
+
+	coverage_profiler.suppressed_assemblies = mono_conc_hashtable_new (g_str_hash, g_str_equal);
+
+	/* Don't need to free content as it is referred to by the lines stored in @filters */
+	content = get_file_content (SUPPRESSION_DIR "/mono-profiler-coverage.suppression");
```

Comment:
Should this use same path and name as a profiler if we want to decouple it from it?

Reply:
We want to be as backward compatible as possible, until we remove the support from the log profiler, so in case we don't have this coverage-specific file, we try to fallback to the log-specific and historic one.